tutorials/epochs/plot_30_epochs_metadata.py

```diff
@@ -0,0 +1,172 @@
+# format query strings. Here's another example:
+
+print(epochs['Concreteness > 6 and WordFrequency < 1'])
+
+###############################################################################
+# Notice that in this dataset, each "condition" (A.K.A., each word) occurs only
+# once, whereas with the :ref:`sample-dataset` dataset each condition (e.g.,
+# "auditory/left", "visual/right", etc) occurred dozens of times. This makes
+# the Pandas querying methods especially useful when you want to aggregate
+# epochs that have different condition names but that share similar stimulus
+# properties. For example, here we group epochs based on the number of letters
+# in the stimulus word, and compare the average signal at electrode ``Pz`` for
+# each group:
+
+evokeds = dict()
+query = 'NumberOfLetters == {}'
+for n_letters in epochs.metadata['NumberOfLetters'].unique():
+    evokeds[str(n_letters)] = epochs[query.format(n_letters)].average()
+
+mne.viz.plot_compare_evokeds(evokeds, cmap=('word length', 'viridis'),
+                             picks='Pz')
+
+###############################################################################
+# We can use a Pandas query string to select specific words for plotting:
+
+words = ['typhoon', 'bungalow', 'colossus', 'drudgery', 'linguist', 'solenoid']
+epochs['WORD in {}'.format(words)].plot(n_channels=29)
+
+###############################################################################
+# Metadata can also be useful for sorting the epochs in an image plot. For
+# example, here we order the epochs based on word frequency to see if there's a
+# pattern to the latency or intensity of the response:
+
+sort_order = np.argsort(epochs.metadata['WordFrequency'])
+epochs.plot_image(order=sort_order, picks='Pz')
+
+###############################################################################
+# Although there's no obvious relationship in this case, such analyses may be
+# useful for metadata variables that more directly index the time course of
+# stimulus processing (such as reaction time).
+#
+# Note also that traditional epochs subselection by condition name still works;
+# MNE-Python will try the traditional method first before falling back on rich
+# metadata querying.
+
+epochs['solenoid'].plot_psd()
+
+###############################################################################
+# Adding metadata to an ``Epochs`` object
+# ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+#
+# You can add a metadata :class:`~pandas.DataFrame` to any
+# :class:`~mne.Epochs` object (or replace existing metadata) simply by
+# assigning to the :attr:`~mne.Epochs.metadata` attribute:
+
+new_metadata = pd.DataFrame(data='foo' * len(epochs), columns=['bar'],
+                            index=range(len(epochs)))
+epochs.metadata = new_metadata
+
+###############################################################################
+# TODO: deleting metadata (is it even possible?)
```

Comment:
Yes, you can just do `epochs.metadata = None`. But there is no `epochs.set_metadata`, and doing `epochs.metadata = new_metadata` is dangerous because it's not checked in any way, so maybe we shouldn't advertise this. Alternatively if you want to refactor `Epochs.__init__` to internally just call `self.set_metadata(metadata)` somewhere, then maybe we can make `epochs.set_metadata(new_metadata)` and `epochs.set_metadata(None)` work without too much extra effort.

Reply:
when I tried `epochs.metadata = None` it did not in fact delete the metadata. So yeah, I think a `set_metadata()` method is needed.

```
In [1]: import os 
   ...: import mne 
   ...:  
   ...: kiloword_data_folder = mne.datasets.kiloword.data_path() 
   ...: kiloword_data_file = os.path.join(kiloword_data_folder, 
   ...:                                   'kword_metadata-epo.fif') 
   ...: epochs = mne.read_epochs(kiloword_data_file) 
   ...:                                                                                                                                            
Reading /home/drmccloy/mne_data/MNE-kiloword-data/kword_metadata-epo.fif ...
Isotrak not found
    Found the data of interest:
        t =    -100.00 ...     920.00 ms
        0 CTF compensation matrices available
960 matching events found
No baseline correction applied
Adding metadata with 8 columns
0 projection items activated
In [2]: epochs.metadata.head()                                                                                                                     
Out[2]: 
   WORD  Concreteness  WordFrequency  OrthographicDistance  NumberOfLetters  BigramFrequency  ConsonantVowelProportion  VisualComplexity
0  film          5.45       3.189490                  1.75              4.0           343.25                      0.75         55.783710
1  cent          5.90       3.700704                  1.35              4.0           546.75                      0.75         63.141553
2  shot          4.60       2.858537                  1.20              4.0           484.75                      0.75         64.600033
3  cold          3.70       3.454540                  1.15              4.0          1095.25                      0.75         63.657457
4  main          3.00       3.539076                  1.35              4.0           686.00                      0.50         68.945661
In [3]: epochs.metadata = None                                                                                                                     
Replacing existing metadata with 8 columns
In [4]: epochs.metadata.head()                                                                                                                     
Out[4]: 
   WORD  Concreteness  WordFrequency  OrthographicDistance  NumberOfLetters  BigramFrequency  ConsonantVowelProportion  VisualComplexity
0  film          5.45       3.189490                  1.75              4.0           343.25                      0.75         55.783710
1  cent          5.90       3.700704                  1.35              4.0           546.75                      0.75         63.141553
2  shot          4.60       2.858537                  1.20              4.0           484.75                      0.75         64.600033
3  cold          3.70       3.454540                  1.15              4.0          1095.25                      0.75         63.657457
4  main          3.00       3.539076                  1.35              4.0           686.00                      0.50         68.945661
In [5]:  

```